examples/app/random_tiles/main.py

```diff
@@ -0,0 +1,39 @@
+from bokeh.plotting import Figure
+from bokeh.io import curdoc
+from bokeh.core.properties import (String, List)
+from bokeh.models.tiles import MercatorTileSource
+from bokeh.tile_providers import STAMEN_TONER
+
+class RandomTileSource(MercatorTileSource):
+    __implementation__ = """
+    _ = require "underscore"
+    Util = require "util/util"
+    MercatorTileSource = require "renderer/tile/mercator_tile_source"
+    class RandomTileSource extends MercatorTileSource
+      type: "RandomTileSource"
+      get_image_url: (x, y, z) ->
+        urls = @get('urls')
+        url = urls[Math.floor(Math.random() * urls.length)]
+        image_url = @string_lookup_replace(url, @get('extra_url_vars'))
+        [x, y, z] = @tms_to_wmts(x, y, z)
+        return image_url.replace("{X}", x).replace('{Y}', y).replace("{Z}", z)
+    module.exports =
+      Model: RandomTileSource
+    """
+    urls = List(String)
```

Comment:
It would be sweet in a follow-on patch if the `__implementation__` could be in `models/random_tile_source.coffee` and the python class in `models/random_tile_source.py` and we would automatically set `__implementation__` based on the matching filenames... shouldn't be too hard right? and then people can use the actual editor mode for the coffeescript instead of editing this big inline string.

For bonus points when doing this, if an app contains `models/` we would always bundle `bokeh-compiler` even if the initial document doesn't contain a custom model (I believe we have a bug right now that we decide on `bokeh-widgets` and `bokeh-compiler` bundling based on the initial doc, so any later changes to the doc can't use widgets/compiler if the original doc didn't?)